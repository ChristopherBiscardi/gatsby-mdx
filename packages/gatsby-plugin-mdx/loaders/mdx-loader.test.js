const mdxLoader = require("./mdx-loader");
const prettier = require("prettier");
const c = require("js-combinatorics");

function genMDXFile(input) {
  const code = {
    frontmatter: `---
one: two
three: 4
array: [1,2,3]
---`,
    defaultLayout: `export default ({children, ...props}) => (
<div>
{children}
</div>
)`,
    namedExports: `export const meta = {author: "chris"}`,
    body: `# Some title
    
a bit of a paragraph
    
some content`
  };

  return {
    name:
      Object.entries(input)
        .filter(([k, v]) => !!v) // eslint-disable-line no-unused-vars
        .map(([k, v]) => k) // eslint-disable-line no-unused-vars
        .join("-") || "body",
    content: [
      input.frontmatter ? code.frontmatter : "",
      input.layout ? code.defaultLayout : "",
      input.namedExports ? code.namedExports : "",
      code.body
    ].join("\n\n")
  };
}

// generate a table of all possible combinations of genMDXfile input
const fixtures = c
  .baseN([true, false], 3)
  .toArray()
  .map(([frontmatter, layout, namedExports]) =>
    genMDXFile({ frontmatter, layout, namedExports })
  )
  .map(({ name, content }) => [
    name,
    {
      internal: { type: "File" },
      sourceInstanceName: "webpack-test-fixtures",
      absolutePath: `/fake/${name}`
    },
    content
  ]);

describe("mdx-loader", () => {
  expect.addSnapshotSerializer({
    print(val /*, serialize */) {
      return prettier.format(val, { parser: "babylon" });
    },
    test() {
      return true;
    }
  });
  test.each(fixtures)(
    "snapshot with %s",
    async (filename, fakeGatsbyNode, content) => {
      const loader = mdxLoader.bind({
        async() {
          return (err, result) => {
            expect(err).toBeNull();
            expect(result).toMatchSnapshot();
          };
        },
        query: {
          getNodes() {
            return fixtures.map(([, node]) => node);
          },
          pluginOptions: {}
        },
        resourcePath: fakeGatsbyNode.absolutePath
      });
      await loader(content);
    }
  );
});
