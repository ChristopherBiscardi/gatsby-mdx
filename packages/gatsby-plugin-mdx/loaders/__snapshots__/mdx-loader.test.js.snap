// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mdx-loader snapshot with body 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);
    this.layout = null;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag name="wrapper" components={components}>
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {};

`;

exports[`mdx-loader snapshot with frontmatter 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);
    this.layout = null;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag name="wrapper" components={components}>
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {
  one: "two",
  three: 4,
  array: [1, 2, 3]
};

`;

exports[`mdx-loader snapshot with frontmatter-layout 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);

    this.layout = ({ children, ...props }) => <div>{children}</div>;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag
        name="wrapper"
        Layout={this.layout}
        layoutProps={props}
        components={components}
      >
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {
  one: "two",
  three: 4,
  array: [1, 2, 3]
};

`;

exports[`mdx-loader snapshot with frontmatter-layout-namedExports 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export const meta = {
  author: "chris"
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);

    this.layout = ({ children, ...props }) => <div>{children}</div>;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag
        name="wrapper"
        Layout={this.layout}
        layoutProps={props}
        components={components}
      >
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {
  one: "two",
  three: 4,
  array: [1, 2, 3]
};

`;

exports[`mdx-loader snapshot with frontmatter-namedExports 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export const meta = {
  author: "chris"
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);
    this.layout = null;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag name="wrapper" components={components}>
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {
  one: "two",
  three: 4,
  array: [1, 2, 3]
};

`;

exports[`mdx-loader snapshot with layout 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);

    this.layout = ({ children, ...props }) => <div>{children}</div>;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag
        name="wrapper"
        Layout={this.layout}
        layoutProps={props}
        components={components}
      >
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {};

`;

exports[`mdx-loader snapshot with layout-namedExports 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export const meta = {
  author: "chris"
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);

    this.layout = ({ children, ...props }) => <div>{children}</div>;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag
        name="wrapper"
        Layout={this.layout}
        layoutProps={props}
        components={components}
      >
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {};

`;

exports[`mdx-loader snapshot with namedExports 1`] = `
import React from "react";
import { MDXTag } from "@mdx-js/tag";

export const meta = {
  author: "chris"
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props);
    this.layout = null;
  }

  render() {
    const { components, ...props } = this.props;
    return (
      <MDXTag name="wrapper" components={components}>
        <MDXTag name="h1" components={components}>{\`Some title\`}</MDXTag>
        <MDXTag
          name="p"
          components={components}
        >{\`a bit of a paragraph\`}</MDXTag>
        <MDXTag name="p" components={components}>{\`some content\`}</MDXTag>
      </MDXTag>
    );
  }
}
export const _frontmatter = {};

`;
